File: TrainingPractice/TrainingPractice/app/src/main/java/com/example/trainingpractice/Data/DatabaseHandler.java
package com.example.trainingpractice.Data;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.Nullable;

import com.example.trainingpractice.Model.Contact;
import com.example.trainingpractice.Utils.Util;

import java.util.ArrayList;
import java.util.List;


public class DatabaseHandler extends SQLiteOpenHelper {


    public DatabaseHandler(@Nullable Context context) {
        super(context, Util.DATABASE_NAME,null, Util.DATABASE_VERSION);
    }

    // we are creating our table in following method
    @Override
    public void onCreate(SQLiteDatabase db) {
        // SQL - Structured Query Language
        String CREATE_CONTACT_TABLE = "CREATE TABLE " + Util.TABLE_NAME + "("
                + Util.KEY_ID + " INTEGER PRIMARY KEY," + Util.KEY_NAME + " TEXT,"
                + Util.KEY_PHONE_NUMBER + " TEXT" + ")";

        db.execSQL(CREATE_CONTACT_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        // deleting table
        db.execSQL("DROP TABLE IF EXISTS " + Util.TABLE_NAME);

        // Create table again

        onCreate(db);
    }

    /**
     * CRUD Operation
     */

    // Add Contact
    public void addContact(Contact contact){
        SQLiteDatabase db  =this.getWritableDatabase();

        ContentValues value = new ContentValues();
        value.put(Util.KEY_NAME, contact.getName());
        value.put(Util.KEY_PHONE_NUMBER,contact.getPhoneNumber());

        //insert to row

        db.insert(Util.TABLE_NAME,null,value);
        db.close();
    }

    // Get a contact

    public Contact getContact(int id){

        SQLiteDatabase db = this.getReadableDatabase();

        Cursor cursor = db.query(Util.TABLE_NAME, new String[] { Util.KEY_ID,
        Util.KEY_NAME,Util.KEY_PHONE_NUMBER},Util.KEY_ID + "=?",
                new String[] {String.valueOf(id)},null,null,null,null);

        if (cursor != null) {
            cursor.moveToFirst();
        }
            Contact contact = new Contact(Integer.parseInt(cursor.getString(0)),
                    cursor.getString(1),cursor.getString(2));

            return contact;
    }

    // Get all contacts

    public List<Contact>  getAllContacts(){
        SQLiteDatabase db = this.getReadableDatabase();

        List<Contact> contactList = new ArrayList<>();

        //select all contacts

        String selectAll = "SELECT * FROM "+Util.TABLE_NAME;
        Cursor cursor = db.rawQuery(selectAll,null);

        // Loop through our contact

        if (cursor.moveToFirst()){
            do {
                Contact contact  =new Contact();
                contact.setId(Integer.parseInt(cursor.getString(0)));
                contact.setName(cursor.getString(1));
                contact.setPhoneNumber(cursor.getString(2));

                // add contact to our list

                contactList.add(contact);
            }while (cursor.moveToNext());
        }
        return contactList;
    }

    //update contact

    public int updateContacts(Contact contact){
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues value = new ContentValues();
        value.put(Util.KEY_NAME,contact.getName());
        value.put(Util.KEY_PHONE_NUMBER,contact.getPhoneNumber());

        // updte row
        return db.update(Util.TABLE_NAME,value,Util.KEY_ID + "=?",
                new String[]{String.valueOf(contact.getId())});

    }

    //Delete single Contact
    public void deleteContact(Contact contact){
        SQLiteDatabase db = this.getWritableDatabase();
        db.delete(Util.TABLE_NAME,Util.KEY_ID + "=?",
                new String[]{String.valueOf(contact.getId())});

        db.close();
    }


    //Get contacts count
    public int getContactCount(){
        String countQuery = "SELECT * FROM "+ Util.TABLE_NAME;
        SQLiteDatabase db = this.getReadableDatabase();
        Cursor cursor = db. rawQuery(countQuery,null);

        // we can't return database after closing database
        //        db.close();

        return cursor.getCount();
    }
}

File: ReadingJson/app/src/main/java/abdulhannanmayo/app/readingjson/Handler/DatabaseHandler.kt
package abdulhannanmayo.app.readingjson.Handler

import abdulhannanmayo.app.readingjson.Model.PersonModelClass
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper

//creating the database logic, extending the SQLiteOpenHelper base class
class DatabaseHandler(context: Context): SQLiteOpenHelper(context,DATABASE_NAME,null,DATABASE_VERSION) {
    companion object {
        private val DATABASE_VERSION = 3
        private val DATABASE_NAME = "EmployeeDatabase"
        private val TABLE_CONTACTS = "ProfessorTable"
        private val KEY_ID = "id"
        private val KEY_NAME = "name"
        private val KEY_EMAIL = "email"
        private val KEY_SUBJECT = "subject"
    }
    override fun onCreate(db: SQLiteDatabase?) {
        // TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        //creating table with fields
        val CREATE_CONTACTS_TABLE = ("CREATE TABLE " + TABLE_CONTACTS + "("
                + KEY_ID + " INTEGER PRIMARY KEY," + KEY_NAME + " TEXT,"+ KEY_EMAIL + " TEXT,"
                + KEY_SUBJECT + " TEXT" + ")")
        db?.execSQL(CREATE_CONTACTS_TABLE)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        //  TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        db!!.execSQL("DROP TABLE IF EXISTS " + TABLE_CONTACTS)
        onCreate(db)
    }


    //method to insert data
    fun addEmployee(emp: PersonModelClass):Long{
        val db = this.writableDatabase
        val contentValues = ContentValues()
        contentValues.put(KEY_ID, emp.userId)
        contentValues.put(KEY_NAME, emp.userName) // EmpModelClass Name
        contentValues.put(KEY_EMAIL,emp.userSurName ) // EmpModelClass Phone
        contentValues.put(KEY_SUBJECT,emp.userSubject ) // EmpModelClass Phone
        // Inserting Row
        val success = db.insert(TABLE_CONTACTS, null, contentValues)
        //2nd argument is String containing nullColumnHack
        db.close() // Closing database connection
        return success
    }
    //method to read data
    fun viewEmployee():List<PersonModelClass>{
        val empList:ArrayList<PersonModelClass> = ArrayList<PersonModelClass>()
        val selectQuery = "SELECT  * FROM $TABLE_CONTACTS"
        val db = this.readableDatabase
        var cursor: Cursor? = null
        try{
            cursor = db.rawQuery(selectQuery, null)
        }catch (e: SQLiteException) {
            db.execSQL(selectQuery)
            return ArrayList()
        }
        var userId: Int
        var userName: String
        var userEmail: String
        var userSubject: String
        if (cursor.moveToFirst()) {
            do {
                userId = cursor.getInt(cursor.getColumnIndex("id"))
                userName = cursor.getString(cursor.getColumnIndex("name"))
                userEmail = cursor.getString(cursor.getColumnIndex("email"))
                userSubject = cursor.getString(cursor.getColumnIndex("subject"))
                val emp= PersonModelClass(userId = userId.toString(), userName = userName, userSurName = userEmail, userSubject = userSubject)
                empList.add(emp)
            } while (cursor.moveToNext())
        }
        return empList
    }

}